rg_instructor_analytics/tasks.py

```diff
@@ -23,3 +36,90 @@ def send_email_to_cohort(subject, message, students):
     msg.encoding = 'UTF-8'
     msg.attach_alternative(html_content, "text/html")
     msg.send(fail_silently=False)
+
+
+cron_settings = getattr(
+    settings, 'RG_ANALYTICS_ENROLLMENT_STAT_UPDATE',
+    {
+        'minute': '*/1',
+        'hour': '*',
+        'day_of_week': '*',
+        'day_of_month': '*',
+        'month_of_year': '*',
+    }
+)
+
+
+@periodic_task(run_every=crontab(**cron_settings))
+def enrollment_collector_date():
+    """
+    Task for update enrollment cache.
+    """
+    last_cachce = EnrollmentByStudent.objects.all().order_by('-last_update')
+    if last_cachce.exists():
+        last_update = last_cachce.first().last_update
+    else:
+        last_update = datetime(2000, 01, 01, 0, 0)
+    enrollments_history = (
+        CourseEnrollment.history
+        .filter(~Q(history_type='+'))
+        .filter(history_date__gt=last_update)
+        .values("history_date", "is_active", "user", "course_id")
+        .order_by('history_date')
+    )
+    users_state = {
+        (u['student'], u['course_id']): {'last_update': u['last_update'], 'state': u['state']}
+        for u in EnrollmentByStudent.objects.all().values("course_id", "student", "last_update", "state", )
+    }
+    total_stat = {
+        t['course_id']: t['total'] for t in TotalEnrollmentByCourse.objects.all().values("course_id", "total", )
+    }
+    result_stat = {}
+    for e in enrollments_history:
+        key = e['user'], e['course_id']
+        if key in users_state and users_state[key]['state'] == e['is_active']:
+            continue
+        users_state[key] = {
+            'last_update': e['history_date'],
+            'state': e['is_active'],
+        }
+        total_key = e['history_date'].date(), e['course_id']
```

Comment:
I would use parentheses with the tuple definition

Reply:
fix